include/map_observer.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include "rust/cxx.h"
+#include <mbgl/map/map_observer.hpp>
+#include <memory>
+#include <optional>
+
+namespace mln {
+namespace bridge {
+
+    // Forward declarations
+    using MapObserverCameraChangeMode = mbgl::MapObserver::CameraChangeMode; // Required, because enum nested in class is not supported by cxx
+
+    class VoidCallback;
+    void void_callback(VoidCallback const& callback) noexcept;
+
+    class FailingLoadingMapCallback;
+    void failing_loading_map_callback(FailingLoadingMapCallback const& callback, mbgl::MapLoadError error, const rust::Str what) noexcept;
+
+    class CameraDidChangeCallback;
+    void camera_did_change_callback(CameraDidChangeCallback const& callback, MapObserverCameraChangeMode mode) noexcept;
+
+    class FinishRenderingFrameCallback;
+    void finish_rendering_frame_callback(FinishRenderingFrameCallback const& callback, bool needsRepaint, bool placementChanged) noexcept;
+
+    class MapObserver: public mbgl::MapObserver {
+        public:
+            void setWillStartLoadingMapCallback(rust::Box<VoidCallback> callback) {
+                willStartLoadingMapCallback = std::optional<rust::Box<VoidCallback>>{std::move(callback)};
+            }
+
+            void setFinishLoadingStyleCallback(rust::Box<VoidCallback> callback) {
+                finishLoadingStyleCallback = std::optional<rust::Box<VoidCallback>>{std::move(callback)};
+            }
+
+            void setBecomeIdleCallback(rust::Box<VoidCallback> callback) {
+                becomeIdleCallback = std::optional<rust::Box<VoidCallback>>{std::move(callback)};
+            }
+
+            void setFailLoadingMapCallback(rust::Box<FailingLoadingMapCallback> callback) {
+                failLoadingMapCallback = std::optional<rust::Box<FailingLoadingMapCallback>>{std::move(callback)};
+            }
+
+            void setFinishRenderingFrameCallback(rust::Box<FinishRenderingFrameCallback> callback) {
+                finishRenderingFrameCallback = std::optional<rust::Box<FinishRenderingFrameCallback>>{std::move(callback)};
+            }
+
+            void setCameraDidChangeCallback(rust::Box<CameraDidChangeCallback> callback) {
+                cameraDidChangeCallback = std::optional<rust::Box<CameraDidChangeCallback>>{std::move(callback)};
+            }
+
+        private:
+            void onWillStartLoadingMap() override {
+                try {
+                    void_callback(*willStartLoadingMapCallback.value());
+                } catch (...) {}
```

Comment:
why do you always ignore exceptions here, should it be at least reported as an error somehow?

Reply:
all callbacks are optional. If the optional is not set and we call .value() an exception will be thrown.
https://en.cppreference.com/w/cpp/utility/optional/value.html
I didn't find a similar function than `inspect` for rust